File: src/store.ts
// Encrypted-at-rest token store.
//
// Persists the OAuth token set to <BRIDGE_HOME>/tokens.json, sealed via
// src/crypto.ts (AES-256-GCM, chmod 600). Never log token material.
import { chmodSync, existsSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { join } from "node:path";
import { ensureBridgeHome, openJson, sealJson, type Sealed } from "./crypto.ts";

export interface TokenSet {
  accessToken: string;
  refreshToken?: string;
  tokenType: string;
  /** Absolute expiry in epoch seconds. */
  expiresAt: number;
  scope: string;
  /** RFC 7638 thumbprint of the DPoP key the token is bound to. */
  jkt: string;
}

export class TokenStore {
  private readonly path: string;

  constructor(private readonly home: string) {
    this.path = join(home, "tokens.json");
  }

  /** Load the persisted token set, or null if none exists. */
  load(): TokenSet | null {
    if (!existsSync(this.path)) return null;
    const sealed = JSON.parse(readFileSync(this.path, "utf8")) as Sealed;
    return openJson<TokenSet>(this.home, sealed);
  }

  /** Persist the token set, encrypted, chmod 600. */
  save(set: TokenSet): void {
    ensureBridgeHome(this.home);
    const sealed = sealJson(this.home, set);
    writeFileSync(this.path, JSON.stringify(sealed), { mode: 0o600 });
    chmodSync(this.path, 0o600);
  }

  /** Remove the persisted token set. */
  clear(): void {
    if (existsSync(this.path)) rmSync(this.path, { force: true });
  }

  /** True if the set is expired (or within `skew` seconds of expiring). */
  isExpired(set: TokenSet, skew = 60): boolean {
    const now = Math.floor(Date.now() / 1000);
    return now >= set.expiresAt - skew;
  }
}

File: internal/store/store.go
// Package store persists the OAuth token set, encrypted at rest, the Go port of
// src/store.ts. The token set is sealed (AES-256-GCM via the seal package) and
// written to <BRIDGE_HOME>/tokens.json with mode 0600. Never log token material.
package store

import (
	"encoding/json"
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"time"

	"github.com/joewitt99/bridge-mcp-client/internal/seal"
)

// DefaultSkew is the expiry safety margin (seconds) used by getAccessToken.
const DefaultSkew int64 = 60

// TokenSet is the persisted OAuth token bundle.
type TokenSet struct {
	AccessToken  string `json:"accessToken"`
	RefreshToken string `json:"refreshToken,omitempty"`
	TokenType    string `json:"tokenType"`
	ExpiresAt    int64  `json:"expiresAt"` // absolute expiry, epoch seconds
	Scope        string `json:"scope"`
	JKT          string `json:"jkt"` // thumbprint of the bound DPoP key
}

// TokenStore reads/writes the encrypted token file under BRIDGE_HOME.
type TokenStore struct {
	home string
	path string
	now  func() time.Time // injectable for deterministic expiry tests
}

// New returns a store rooted at home.
func New(home string) *TokenStore {
	return &TokenStore{home: home, path: filepath.Join(home, "tokens.json"), now: time.Now}
}

// Load returns the persisted token set, or (nil, nil) if none exists.
func (s *TokenStore) Load() (*TokenSet, error) {
	b, err := os.ReadFile(s.path)
	if errors.Is(err, fs.ErrNotExist) {
		return nil, nil
	}
	if err != nil {
		return nil, err
	}
	var sealed seal.Sealed
	if err := json.Unmarshal(b, &sealed); err != nil {
		return nil, err
	}
	var ts TokenSet
	if err := seal.OpenJSON(s.home, sealed, &ts); err != nil {
		return nil, err
	}
	return &ts, nil
}

// Save persists the token set, encrypted, mode 0600.
func (s *TokenStore) Save(ts TokenSet) error {
	if err := seal.EnsureBridgeHome(s.home); err != nil {
		return err
	}
	sealed, err := seal.SealJSON(s.home, ts)
	if err != nil {
		return err
	}
	data, err := json.Marshal(sealed)
	if err != nil {
		return err
	}
	if err := os.WriteFile(s.path, data, 0o600); err != nil {
		return err
	}
	return os.Chmod(s.path, 0o600)
}

// Clear removes the persisted token set (idempotent).
func (s *TokenStore) Clear() error {
	err := os.Remove(s.path)
	if errors.Is(err, fs.ErrNotExist) {
		return nil
	}
	return err
}

// IsExpired reports whether the set is expired or within skew seconds of expiry.
func (s *TokenStore) IsExpired(ts TokenSet, skew int64) bool {
	return s.now().Unix() >= ts.ExpiresAt-skew
}
